plugins/processors/snmp_lookup/lookup_test.go

```diff
@@ -0,0 +1,622 @@
+		require.EqualValues(t, 1, tsc.calls.Load())
+	})
+
+	t.Run("connection fail", func(t *testing.T) {
+		p.getConnectionFunc = func(string) (snmp.Connection, error) {
+			return nil, errors.New("Random connection error")
+		}
+
+		require.Nil(t, p.updateAgent("127.0.0.1"))
+	})
+}
+
+func TestAdd(t *testing.T) {
+	tests := []struct {
+		name     string
+		input    telegraf.Metric
+		expected []telegraf.Metric
+	}{
+		{
+			name:     "no source tag",
+			input:    testutil.MockMetrics()[0],
+			expected: testutil.MockMetrics(),
+		},
+		{
+			name: "no index tag",
+			input: testutil.MustMetric(
+				"test",
+				map[string]string{
+					"source": "127.0.0.1",
+				},
+				map[string]interface{}{"value": 42},
+				time.Unix(0, 0),
+			),
+			expected: []telegraf.Metric{
+				testutil.MustMetric(
+					"test",
+					map[string]string{
+						"source": "127.0.0.1",
+					},
+					map[string]interface{}{"value": 42},
+					time.Unix(0, 0),
+				),
+			},
+		},
+		{
+			name: "cached",
+			input: testutil.MustMetric(
+				"test",
+				map[string]string{
+					"source": "127.0.0.1",
+					"index":  "123",
+				},
+				map[string]interface{}{"value": 42},
+				time.Unix(0, 0),
+			),
+			expected: []telegraf.Metric{
+				testutil.MustMetric(
+					"test",
+					map[string]string{
+						"source": "127.0.0.1",
+						"index":  "123",
+						"ifName": "eth123",
+					},
+					map[string]interface{}{"value": 42},
+					time.Unix(0, 0),
+				),
+			},
+		},
+		{
+			name: "non-existing index",
+			input: testutil.MustMetric(
+				"test",
+				map[string]string{
+					"source": "127.0.0.1",
+					"index":  "999",
+				},
+				map[string]interface{}{"value": 42},
+				time.Unix(0, 0),
+			),
+			expected: []telegraf.Metric{
+				testutil.MustMetric(
+					"test",
+					map[string]string{
+						"source": "127.0.0.1",
+						"index":  "999",
+					},
+					map[string]interface{}{"value": 42},
+					time.Unix(0, 0),
+				),
+			},
+		},
+	}
+
+	tsc := &testSNMPConnection{}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			plugin := Lookup{
+				AgentTag:        "source",
+				IndexTag:        "index",
+				ClientConfig:    *snmp.DefaultClientConfig(),
+				CacheSize:       defaultCacheSize,
+				CacheTTL:        defaultCacheTTL,
+				ParallelLookups: defaultParallelLookups,
+				Log:             testutil.Logger{Name: "processors.snmp_lookup"},
+			}
+			require.NoError(t, plugin.Init())
+
+			var acc testutil.Accumulator
+			require.NoError(t, plugin.Start(&acc))
+			defer plugin.Stop()
+
+			plugin.getConnectionFunc = func(string) (snmp.Connection, error) {
+				return tsc, nil
+			}
+
+			// Sneak in cached  data
+			plugin.cache.cache.Add("127.0.0.1", &tagMap{rows: map[string]map[string]string{"123": {"ifName": "eth123"}}})
+
+			// Do the testing
+			require.NoError(t, plugin.Add(tt.input, &acc))
+			require.Eventually(t, func() bool {
+				return int(acc.NMetrics()) >= len(tt.expected)
+			}, 3*time.Second, 100*time.Millisecond)
```

Comment:
@srebhan Can't we replace these with this?
```suggestion
			acc.Wait(len(tt.expected))
```

Reply:
No! `acc.Wait` waits forever if this fails and you get a CircleCI timeout after 10min. This introduces cost if it ever fails. So we rather should get rid of `acc.Wait` as `require.Eventually` is really the right way of doing this!

Reply:
See my edit, it is proposed to be used and also used a lot in tests. If it shouldn't then that recommendation should be changed.

The test accumulator should maybe have a `Eventually` and `EventuallyError` counterpart of `Wait` and `WaitError`.
```go
func (a *Accumulator) Eventually(t *testing.T, n int, waitFor time.Duration, tick time.Duration, msgAndArgs ...interface{}) {
	t.Helper()

	require.Eventually(t, func() bool {
		return int(a.NMetrics()) >= n
	}, waitFor, tick, msgAndArgs...)
}
```

Then this test could be changed to:
```suggestion
			acc.Eventually(t, len(tt.expected), 3*time.Second, 100*time.Millisecond)
```